### PR TITLE
Add metadata to device

### PR DESCRIPTION
**Pull request by matt-mazzucato**
Address #145

### CODE REVIEW

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex [hunk 1]
```diff
@@ -437,7 +437,7 @@ defmodule Astarte.AppEngine.API.Device.Queries do
         device_id,
         %InterfaceDescriptor{storage_type: :multi_interface_individual_datastream_dbtable} =
           interface_descriptor,
-        endpoint_id,
+        _endpoint_id,
```
Comment (bettio) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex]: I think that we should fix any warning in release-0.11, just change code related to this PR.

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex [hunk 2]
```diff
@@ -847,6 +852,59 @@ defmodule Astarte.AppEngine.API.Device.Queries do
...
+      |> DatabaseQuery.put(:metadata_value, metadata_value)
+      |> DatabaseQuery.put(:device_id, device_id)
+      |> DatabaseQuery.consistency(:each_quorum)
+
+    with {:ok, _result} <- DatabaseQuery.call(client, query) do
+      :ok
+    else
+      error ->
+        Logger.warn(
+          "An error occurred while inserting metadata into the devices' table: #{inspect(error)}",
```
Comment (bettio) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex]: Full stop at the end of message.

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex [hunk 3]
```diff
@@ -847,6 +852,59 @@ defmodule Astarte.AppEngine.API.Device.Queries do
...
+      DatabaseQuery.new()
+      |> DatabaseQuery.statement(delete_metadata_statement)
+      |> DatabaseQuery.put(:metadata_key, metadata_key)
+      |> DatabaseQuery.put(:device_id, device_id)
+
+    with {:ok, _result} <- DatabaseQuery.call(client, delete_metadata_query) do
+      :ok
+    else
+      error ->
+        Logger.warn("Delete metadata failed with error: #{inspect(error)}",
```
Comment (bettio) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex]: Full stop at the end of the message

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/device_status.ex [hunk 4]
```diff
@@ -34,6 +34,7 @@ defmodule Astarte.AppEngine.API.Device.DeviceStatus do
     field :first_credentials_request, :utc_datetime
     field :last_credentials_request_ip
     field :last_seen_ip
+    field :metadata, :map
```
Comment (bettio) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/device_status.ex]: map type is quite permissive, while we require maps here to be %{string.t() => string.t()}.
@rbino should we use an ecto changeset just for it?

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/device.ex [hunk 5]
```diff
@@ -101,6 +131,12 @@ defmodule Astarte.AppEngine.API.Device do
     |> Enum.into(%{})
   end
 
+  defp merge_metadata(old_metadata, new_metadata) when is_map(old_metadata) do
```
Comment (bettio) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/device.ex]: do we have a similar function for aliases?

Comment (matt-mazzucato) [reply apps/astarte_appengine_api/lib/astarte_appengine_api/device/device.ex]: Yes. The two functions are now condensed in a more general `merge_data`

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex [hunk 6]
```diff
@@ -847,6 +852,87 @@ defmodule Astarte.AppEngine.API.Device.Queries do
...
+    WHERE device_id = :device_id
+    """
+
+    query =
+      DatabaseQuery.new()
+      |> DatabaseQuery.statement(insert_metadata_statement)
+      |> DatabaseQuery.put(:metadata_key, metadata_key)
+      |> DatabaseQuery.put(:metadata_value, metadata_value)
+      |> DatabaseQuery.put(:device_id, device_id)
+      |> DatabaseQuery.consistency(:each_quorum)
```
Comment (rbino) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex]: If this is `each_quorum`, also the DELETE query should be `:each_quorum`. What do you think about these consistency levels @bettio?

File: apps/astarte_housekeeping/mix.lock [hunk 7]
```diff
@@ -1,7 +1,7 @@
 %{
```
Comment (rbino) [inline apps/astarte_housekeeping/mix.lock]: Why all those deps got updated?

File: apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex [hunk 8]
```diff
@@ -741,7 +744,9 @@ defmodule Astarte.AppEngine.API.Device.Queries do
 
     with {:ok, result} <- DatabaseQuery.call(client, device_query),
          device_row when is_list(device_row) <- DatabaseResult.head(result) do
-      {:ok, device_status_row_to_device_status(device_row)}
+      device_status = device_status_row_to_device_status(device_row)
+
+      {:ok, device_status}
```
Comment (rbino) [inline apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex]: Unnecessary change

File: apps/astarte_appengine_api/mix.lock [hunk 9]
```diff
@@ -1,7 +1,7 @@
 %{
```
Comment (rbino) [inline apps/astarte_appengine_api/mix.lock]: Why these deps got updated?

File: CHANGELOG.md [hunk 10]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [Unreleased]
 ### Changed
 - [appengine_api] Always return an object when GETting on object aggregated interfaces.
+### Added
```
Comment (rbino) [inline CHANGELOG.md]: Needs an extra whitespace above
